### PR TITLE
Add code that changes the faction of Tortheldrin in Dire Maul

### PR DESCRIPTION
**Pull request by ASBaj**
**Changes proposed:**
- Per WoW Classic, Tortheldrin becomes hostile after Immolthar is killed. This code implements an event that does this.

**Issues addressed:**

Closes (insert issue tracker number)

**Tests performed:**

- Code builds, tested that Torheldrin becomes hostile after Immolthar dies.

**Known issues and TODO list:** (add/remove lines as needed)

- None

### CODE REVIEW

**Comment by ASBaj** [comment]
Discovered a bug. Didn't do null checks, so server crashes if re-entering the instance after the bosses have been killed. Will fix.

**Comment by ASBaj** [comment]
Fixed bug in 285330422315b453bb760c18f7de046f34143aac.

**Comment by iCry1337** [comment]
Shauren / Jackpotz Gaylords

File: instance_dire_maul.cpp [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        }
+
+        void OnCreatureCreate(Creature* creature) override
+        {
+            switch (creature->GetEntry())
+            {
+                case NPC_IMMOLTHAR:
+                    _immoGUID = creature->GetGUID();
+                    // we make Immolthar non attackable, otherwise players with pets can pull him out of the forcefield
+                    // TODO: this change isnt correct but since of today (13.09.2020) mmaps dont support doors
```
Comment (jackpoz) [inline instance_dire_maul.cpp:93]: was this code copied from somewhere ?

File: instance_dire_maul.cpp [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                // remove previously set non attackable flag
+                if (Creature* immo = instance->GetCreature(_immoGUID))
+                    immo->RemoveUnitFlag(UNIT_FLAG_NON_ATTACKABLE);
+            }
+        }
+
+        bool IsImmoDead()
+        {
+            Creature* immo = instance->GetCreature(_immoGUID);
+            if (immo != nullptr && !immo->IsAlive())
```
Comment (jackpoz) [inline instance_dire_maul.cpp:285]:
```suggestion
            if (immo && !immo->IsAlive())

```

File: instance_dire_maul.cpp [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+            {
+                Creature* tortheldrin = instance->GetCreature(_tortheldrinGUID);
+                if (tortheldrin != nullptr)
+                    tortheldrin->SetFaction(FACTION_ENEMY);
```
Comment (jackpoz) [inline instance_dire_maul.cpp:289]:
```suggestion
                if (Creature* tortheldrin = instance->GetCreature(_tortheldrinGUID))
                    tortheldrin->SetFaction(FACTION_ENEMY);

```

File: instance_dire_maul.cpp [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+                return true;
+            }
+            else if (immo == nullptr)
```
Comment (jackpoz) [inline instance_dire_maul.cpp:292]:
```suggestion
            else if (!immo)

```

**Comment by Jinnaix** [comment]
@ASBaj 
You seem to have made a mistake with the location of your file - its in your root folder instead of src/server/scripts/Kalimdor/DireMaul/

Please relocate it, otherwise we are not able to merge this.

**Comment by eXPeRi91** [comment]
`#include "ScriptMgr.h"
#include "World.h"

class TortheldrinFactionChangeEvent : public WorldScript
{
public:
    TortheldrinFactionChangeEvent() : WorldScript("tortheldrin_faction_change_event") { }

    void OnBeforeConfigLoad(bool reload) override
    {
        sConfigMgr->AddWorldScript(WORLD_SCRIPT_EVENT_ON_KILL_CREATURE, "TortheldrinFactionChangeEvent", "OnKillCreature");
    }

    void OnKillCreature(Player* player, Creature* creature)
    {
        if (creature->GetEntry() == 15362) // Immolthar
        {
            // Get Tortheldrin and change its faction to hostile
            if (Creature* tortheldrin = player->GetMap()->GetCreature(player->GetMap()->GetCreatureGUID(17942)))
                tortheldrin->setFaction(FACTION_HOSTILE);
        }
    }
};

void AddSC_tortheldrin_faction_change_event()
{
    new TortheldrinFactionChangeEvent();
}`
I think this will work?

**Comment by ASBaj** [comment]
Sorry all, time is hard to come by, and I forgot about this request.

I would like to get it finished. It looks like others have also made contributions.

What precisely needs to be done to finish this one up? Thanks!

**Comment by Aokromes** [comment]
> Sorry all, time is hard to come by, and I forgot about this request.
> 
> I would like to get it finished. It looks like others have also made contributions.
> 
> What precisely needs to be done to finish this one up? Thanks!

move file to proper directory so it diffs properly and do what jackpoz told you to do.

File: instance_dire_maul.cpp (hunk 1 above)
Comment (ASBaj) [reply instance_dire_maul.cpp]: I don't believe so, that's the original code from the unchanged file.

File: instance_dire_maul.cpp (hunk 2 above)
Comment (ASBaj) [reply instance_dire_maul.cpp]: Done as of f0530402887f2e91769d47adb45736bca45809e7.

File: instance_dire_maul.cpp (hunk 3 above)
Comment (ASBaj) [reply instance_dire_maul.cpp]: Done as of f0530402887f2e91769d47adb45736bca45809e7.

File: instance_dire_maul.cpp (hunk 4 above)
Comment (ASBaj) [reply instance_dire_maul.cpp]: Done as of f0530402887f2e91769d47adb45736bca45809e7.

**Comment by ASBaj** [comment]
@Aokromes @jackpoz I believe all tasks have been completed. Please let me know if there's anything else. Thanks! And sorry for the delay.

**Comment by Aokromes** [comment]
remove added whitespaces :)

**Comment by ASBaj** [comment]
@Aokromes Can you be more specific? I grabbed a copy of the raw, original file and compared it against the content of this pull request, and Git is not showing any added whitespace.

**Comment by heyitsbench** [comment]
https://github.com/TrinityCore/TrinityCore/pull/28581/files

**Comment by Aokromes** [comment]
> @Aokromes Can you be more specific? I grabbed a copy of the raw, original file and compared it against the content of this pull request, and Git is not showing any added whitespace.

compare:
https://github.com/TrinityCore/TrinityCore/pull/28581/files?w=1
https://github.com/TrinityCore/TrinityCore/pull/28581/files

**Comment by ASBaj** [comment]
@Aokromes Very unhelpfully, my Visual Studio was auto formatting text when I pasted it in. That's why the whitespace changes in the first place. I turned it off and recommitted without the whitespace changes. Please let me know if there's anything else!

**Comment by Ovahlord** [comment]
you should use the OnUnitDeath instance script hook to trigger the faction change instead of checking the unit every second. This way, all of this only fires once.

**Comment by ASBaj** [comment]
@Ovahlord @Aokromes Done, OnUnitDeath is now in use. Script has been tested and works correctly.

**Comment by Ovahlord** [comment]
@CraftedRO mind to validate this and test it? Gonna merge it when everything's fine.

**Comment by Aokromes** [comment]
@ASBaj

**Comment by CraftedRO** [comment]
I guess we better remove the Issues addressed and we merge as it is since it does fix the purposed issue, and leave that one with chest for another pr.

**Comment by jackpoz** [comment]
Thanks for the PR 🚀 😃

**Pull request merged**